File: VB/HeckYaBaby/Backend/Infrastructure/Repository.vb
Imports System.Data.Entity
Imports System.Linq.Expressions
Imports Backend.Domain
Imports EntityFramework.Extensions

Public Class Repository
    Implements IRepository

    Private ReadOnly _context As DbContext

    Sub New(context As DbContext)
        _context = context
    End Sub

    Public Sub Add (Of T As Class)(item As T) Implements IRepository.Add
        _context.Set(Of T).Add(item)
    End Sub

    Public Function AllInstances (Of T As Class)() As IQueryable(Of T) Implements IRepository.AllInstances
        return _context.Set(Of T)
    End Function

    Public Function Find (Of T As Class)(predicate As Expression(Of Func(Of T,Boolean))) As T Implements IRepository.Find
        return _context.Set(Of T).FirstOrDefault(predicate)
    End Function

    Public Function NewUnitOfWork() As IUnitOfWork Implements IRepository.NewUnitOfWork
        return new UnitOfWork(_context)
    End Function


    Public Sub Remove (Of T As Class)(predicate As Expression(Of Func(Of T,Boolean))) Implements IRepository.Remove
        _context.Set(Of T).
            Where(predicate).
            Delete()
    End Sub

    Public Sub Remove (Of T As Class)(item As T) Implements IRepository.Remove
        _context.Set(Of T).Remove(item)
    End Sub

    Public Function Update (Of T As Class)(predicate As Expression(Of Func(Of T,Boolean)), map As Expression(Of Func(Of T,T))) As Integer Implements IRepository.Update
        return _context.Set(Of T).
            Where(predicate).
            Update(map)
    End Function

    Private Class UnitOfWork
        Implements IUnitOfWork

        Private ReadOnly _context As DbContext

        Public Sub New(context As DbContext)
            _context = context
        End Sub

        public Sub Commit() Implements IUnitOfWork.Commit
            _context.SaveChanges()
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
        End Sub
    End Class
End Class

File: C#/HeckYaBaby/Service/Infrastructure/Repository.cs
using System;
using System.Data.Entity;
using System.Linq;
using System.Linq.Expressions;
using EntityFramework.Extensions;
using Service.Domain;

namespace Service.Infrastructure
{
    internal class Repository : IRepository 
    {
        private readonly DbContext _context;

        public Repository(DbContext context)
        {
            _context = context;
        }

        public IQueryable<T> AllInstances<T>() where T : class
        {
            return _context.Set<T>();
        }

        public T Get<T>(Expression<Func<T, bool>> predicate) where T : class
        {
            return _context.Set<T>().FirstOrDefault(predicate);
        }

        public void Add<T>(T item) where T : class
        {
            _context.Set<T>().Add(item);
        }

        public void Remove<T>(Expression<Func<T, bool>> predicate) where T : class
        {
            _context.Set<T>()
                .Where(predicate)
                .Delete();
        }

        public void Remove<T>(T item) where T : class
        {
            _context.Set<T>().Remove(item);
        }

        public int Update<T>(Expression<Func<T, bool>> predicate, Expression<Func<T, T>> map) where T : class
        {
            return _context.Set<T>()
                .Where(predicate)
                .Update(map);
        }

        public IUnitOfWork NewUnitOfWork()
        {
            return new UnitOfWork(_context);
        }

        private class UnitOfWork : IUnitOfWork
        {
            private readonly DbContext _context;

            public UnitOfWork(DbContext context)
            {
                _context = context;
            }

            public void Commit()
            {
                _context.SaveChanges();
            }

            public void Dispose()
            {
            }
        }

    }
}